### DISCUSSION TITLE
关于xmake管理同一版本不同配置包的一些疑惑

### DISCUSSION THREAD

**Discussion by xiaomi2077** [xmake-io/xmake#2465] [General]
1、在xmake.lua里写`add_requires("opencv")`和写`add_requires("opencv",configs={gtk=true})`是不是会在`~/.xmake/package/o/opencv/4.5.5/`目录下产生2个目录，分别对应着带gtk和不带gtk版本的opencv
2、在别的项目的xmake.lua中，引用带gtk的opencv，是不是要写成`add_requires("opencv",configs={gtk=true})`才能引用的到带gtk的opencv？`target('xxx')`里的`add_packages("opencv")`链接的也是带有gtk的opencv的lib？
（以我做的实验是这样，想确认下我想的对不对）

**Comment by xiaomi2077** [comment]
当存在`同一版本多个不同配置`的opencv时，如果写才能在项目中引用到自己想要的配置的opencv？（好像这样总结更清楚）

**Reply by waruqi** [reply]
```lua
add_requires("opencv")
add_requires("opencv~gtk",configs={gtk=true})

target("")
    add_packages("opencv")

target("")
    add_packages("opencv~gtk")
```

用 `packagename~xxx` 区分下

参考：https://github.com/xmake-io/xmake/blob/master/tests/projects/package/multiconfig/xmake.lua

**Comment by xiaomi2077** [comment]
1、xmake是如何搜索到相应配置的opencv的？
2、是否不同配置的opencv安装在唯一的以校验码为目录名的目录下？如果是的话，我把目录名重命名下，是否xmake就认为这个配置的xmake不存在了，再次执行xmake时就要再次下载安装？

**Reply by waruqi** [reply]
> 1、xmake是如何搜索到相应配置的opencv的？

过程很复杂，一两句说不清，反正默认优先找系统，找不到，到对应 configs 包下找

2、是否不同配置的opencv安装在唯一的以校验码为目录名的目录下？如果是的话，我把目录名重命名下，是否xmake就认为这个配置的xmake不存在了，再次执行xmake时就要再次下载安装？

不同配置，plat/arch/toolchains/ 混合计算唯一 buildhash 。。buildhash不同，就是不同的包，要重装

**Reply by xiaomi2077** [reply]
xmake是否根据buildhash的目录是否存在判断包是否安装？
比如如果我把`~/.xmake/packages/o/opencv/4.5.5/4709da9c5bcb484baaa87dce9694a753/`中的`4709da9c5bcb484baaa87dce9694a753`改为`TMP_4709da9c5bcb484baaa87dce9694a753`，那在项目中如果引用opencv，是否判断包不存在，重新安装？

**Reply by waruqi** [reply]
包装在 4709da9c5bcb484baaa87dce9694a753 这个目录下，当然从这里面找了，你把它 rename 了，还怎么找到呢？
